Correct handeling of Matrices, ExprCondPair and Boolean expression in sympy.core.function.nfloat()

molysgaard opened the pull request:
#### Brief description of what is fixed or changed
When given a subclass of MatrixBase, an ExprCondPair or a Boolean expression nfloat() now should do the right thing.

#### Other comments
None

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* core
  * sympy.core.function.nfloat() now handles ExprCondPair expressions, Boolean expressions and expressions that are subclasses of MatrixBase.
<!-- END RELEASE NOTES -->

molysgaard committed:
function.py: nfloat(): Correctly handle matrices

sylee957 commented:
Besides the merge conflict that needs to be resolved,

I think that it should't rely on the mutation of the matrix
Because it doesn't work with an immutable matrix
```python
from sympy import *

m = ImmutableMatrix([[1, 2], [3, 4]])
nfloat(m)
```

I'd suggest using `expr.applyfunc(lambda x: nfloat(x))` instead.

sylee957 commented:
This now have merge conflicts.

molysgaard commented:
Rebased onto current `master`.
The merge conflicts contained more general fixes for `ExprCondPair` and `Boolean` so I removed my handling of them. What remains is handling of `MatrixBase`.

sylee957 committed:
Add tests for nfloat for matrices

sylee957 commented:
I see there is no difference in handling `ExprCondPair` after the changes were dropped after the rebase.

```
from sympy import *

x = Symbol('x')
b = nfloat(Piecewise([1, x > 1], [0, True]))
b
```
![image](https://user-images.githubusercontent.com/34944973/66551717-f8906800-eb82-11e9-92ae-03fd59eac98b.png)

I thought that your previous intent was to make `nfloat` to behave for `ExprCondPair` and `Boolean` not to cast the numeric types used in the comparators to float. Was it right?

oscarbenjamin commented:
I guess there were other changes here before the rebase. What is here is just changes to `nfloat(Matrix)` but they look good.

I'll merge this. If other changes are needed they can come in another PR.

molysgaard commented:
> I guess there were other changes here before the rebase.

That is correct. My fixes were based of `sympy-1.4` for which I experienced problems for `Boolean`s, `ExprCondPair` and instances of `MatrixBase`. I first implemented fixes for these on top of `sympy-1.4`, and then when I rebased to `master`, two of these three fixes were already implemented. Because of this the only change left was handling of instances of `MatrixBase`.
